lib/networkaccessmanager.cpp

```diff
@@ -56,7 +60,26 @@ NetworkAccessManager::~NetworkAccessManager() = default;
 QNetworkReply* NetworkAccessManager::createRequest(
     Operation op, const QNetworkRequest& request, QIODevice* outgoingData)
 {
+    if(request.url().scheme() == QStringLiteral("mxc")) {
+        auto url = request.url().toString().mid(6);
+        auto pair = d->images[url];
+        d->images.remove(url);
+        return new MxcReply(pair.first, pair.second, url);
+    }
     auto reply = QNetworkAccessManager::createRequest(op, request, outgoingData);
     reply->ignoreSslErrors(d->ignoredSslErrors);
     return reply;
 }
+
+QStringList NetworkAccessManager::supportedSchemesImplementation() const
+ {
+    auto schemes = QNetworkAccessManager::supportedSchemesImplementation();
+    schemes += QStringLiteral("mxc");
+    return schemes;
+ }
+
+QUrl NetworkAccessManager::load(Room *room, const QString &eventId, const QString &mediaId)
+{
+    d->images[mediaId] = qMakePair(room, eventId);
```

Comment:
That assumes two things at once, both not necessarily correct:
1. Any `NAM::load()` invocation only occurs in the context of a particular event. Room and user avatars are not bound to events (and user avatars are not even bound to rooms).
2. Even under the first assumption, each `mxc` URL can only occur in one event. It is indeed commonplace in today's Matrix but there's really nothing preventing the same `mxc` URL from appearing in two events or even rooms. Quotient, for one, has uploading and message sending pretty much uncoupled.

What I would probably recommend is, as discussed in the chat room, stuffing the account name into the URL's query (or even the fragment - we really don't need to send it anywhere!) so that `createRequest()` could find the appropriate connection object using the freshly introduced `AccountRegistry`. Then you won't need to remember things in a hashmap, the whole thing becomes stateless (and it's a huge benefit in a multithreaded context, which is a case of `NAM`).

Reply:
Right, we need a way of loading an image without an event... I guess that means implementing two different "url schemes", since we definitely need the event for encrypted images/files

Reply:
One can extend the other, no? Also, what about storing the room and event in the URL query or even fragment?

Reply:
Sure, that's what i mean